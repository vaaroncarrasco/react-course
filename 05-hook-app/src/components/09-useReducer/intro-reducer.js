
const initialState = [{
    id: 1,
    todo: 'Comprar frijoles',
    done: false
}];

const todoReducer = (state = initialState, action) => {

    if (action?.type === 'agregar') {
        return [ ...state, action.payload ];
    } // reducer can not have else/return

    return state;
}

let todos = todoReducer();

// ! DO NOT USE .push() IN REACT // DO NOT MUTATE STATE -> state++
// ? ALWAYS CREATE NEW STATE -> state + 1


const newTodo = {
    id: 2,
    todo: 'Comprar arroz',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo,
}

todos = todoReducer( todos, agregarTodoAction );

console.log(todos);