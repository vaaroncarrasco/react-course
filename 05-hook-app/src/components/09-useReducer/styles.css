* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

body {
    background: #141414;
    color: #f8f9f4;
    padding: 4rem;
}

p {
    cursor: pointer;
    margin: 0px;
}

.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.complete {
    text-decoration: line-through;
}