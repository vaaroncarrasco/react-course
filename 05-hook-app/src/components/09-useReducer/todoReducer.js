export const todoReducer = ( state = [], action ) => {

    switch ( action.type ) {
        case 'add':
            return [ ...state, action.payload];

        case 'delete':
            // * return new array without the deleted todo
            return state.filter( todo => todo.id !== action.payload); // 123445

        case 'toggle':
            return state.map( todo =>
                ( todo.id === action.payload ) // if
                    ? { ...todo, done: !todo.done } // return
                    : todo // otherwise return
            )

        case 'toggle-old':
            return state.map( todo => {

                if ( todo.id === action.payload ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }

            });

        default:
            return state;
    }

}