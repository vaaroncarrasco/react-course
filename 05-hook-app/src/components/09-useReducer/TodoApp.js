import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer'; // reducers are imported
import { TodoList } from './components/TodoList';
import { TodoAdd } from './components/TodoAdd';

import './styles.css';

// destructured arrays [ ordered, props ]
// destructured objects { props, unordered:name }

const init = () => { // * returns initial state when calling init

    return JSON.parse(localStorage.getItem('todos')) || []; // if todos is null return []

    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Learn React',
    //     done: false
    // }];
}

// * init sets initialState which is immediatly saved in localStorage w/ 'todos' id as [] or [{}...]
// it rewrites the empty array [] because of using the same id

export const TodoApp = () => { // ? Remember react re-runs component code logic on every change

    const [ todos, dispatch ] = useReducer(todoReducer, [], init); // init runs to initiate initialState

    // ? localStorage saving when todos change - useEffect()
    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify( todos )); // localStorage only saves strings -> JSON stringify/parse
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId
        }
        dispatch( action );
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo  = ( newTodo ) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    }

    return (
        <>
            <h1>TodoApp ({ todos.length }) </h1>
            <hr />

            <div className="row">

                <div className="col-7">
                    <TodoList
                        todos={ todos }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />
                </div>

                <div className="col-5">

                    <TodoAdd
                        handleAddTodo={ handleAddTodo }
                    />

                </div>

            </div>

        </>
    )
}
