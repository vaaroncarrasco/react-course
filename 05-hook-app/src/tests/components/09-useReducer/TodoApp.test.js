import React from 'react';
import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../components/09-useReducer/TodoApp";
import { demoTodos } from '../fixtures/demoTodos';
import { act } from '@testing-library/react';

describe('Testing <TodoApp /> app', () => {

    const wrapper = shallow( <TodoApp />) ;

    // ? Mock localStorage setting/adding/storing items functionality->
    Storage.prototype.setItem = jest.fn(()=>{});

    test('match snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should add todo', () => {

        // * mount() -> for testing whole App in context // shallow() for components basic test
        const wrapper = mount(<TodoApp />); // rendered in a deeper level

        // Running functions
        act( () => {
            // find <TodoAdd /> component handleAddTodo fn prop and run it passing argument
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
        });

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp (2)');
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);

    });

    test('should delete todo', () => {

        wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] ); // find fn and run it .prop('fn')(param);
        wrapper.find('TodoList').prop('handleDelete')( demoTodos[0].id ); // find fn and run it .prop('fn')(param);

        expect( wrapper.find('h1').text().trim() ).toBe('TodoApp (0)');

    });

});