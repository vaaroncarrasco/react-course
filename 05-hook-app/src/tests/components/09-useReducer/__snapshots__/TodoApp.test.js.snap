// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <TodoApp /> app match snapshot 1`] = `
<Fragment>
  <h1>
    TodoApp (
    0
    ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
